### DISCUSSION TITLE
File system performance improvement

### DISCUSSION THREAD

**Discussion by tonycody** [microsoft/WSL#9068] [General]
https://github.com/MicrosoftDocs/WSL/issues/604

Please check the issue.

One of the great advantages of WSL2 is the seamless use of Linux on Windows and the interworking of file systems.

It is of great significance for software developers. 
It also makes up for some troublesome things of developing software on Windows. 
For example, some operations can be more convenient under Linux. 
This can be achieved through the WSL2 subsystem. 
But I can't have the same code on Windows, in the WSL2 subsystem, which obviously doesn't reflect the value that WSL2 brings to the Windows platform.

If the code needed to build the Java project in Linux is in the Windows System D drive, the performance at build time is very poor. 
The code of the Windows System D drive must be copied to the Linux file system. 
If so, why should I use the WSL2 Linux subsystem? 
Instead of using VirtualBox? 
If WSL2's performance for accessing the Windows file system is extremely poor, why do you want to exchange visits? 
In this way, WSL2 seems to be of little value!

**Comment by xiaokentrl** [comment]
### I'm also a Java developer
### When I was using WSL2, I wasted about a month. My conclusion is:
**1.  WSL 's file system does not mount as well as VirtualBox does.**
**2.  The code is saved in the user directory of disk C, which is very unsafe and inconvenient to set up the WSL installation disk. If you write Java or PHP code, it is recommended to buy a Mac computer.**

**Comment by WesleyBlancoYuan** [comment]
I still find the value of WSL 2 even I cannot create hard links of dirs between file systems (`ln` gives errors of `Invalid cross-device link`) and I have to copy. Source code dirs are relatively small(they should be; I will try to avoid copy `targets` or other binaries-containing dirs); the effort of copying is too small comparing to dual-boot/launching VirtualBox. At last, Ubuntu 20 without X server running on my Win 10 with only 16 MB of memory occupation and I have access to all shortcuts I created in my `.bashrc` and I don't have to memorise the clunky cmd functions/syntax, even for a simple `for` loop. I cannot ask for more. 

And it's only slow when Maven building. For other scenarios it's fine. 

If copying is too annoying, consider using `rsync` to sync files between dirs. Use `git ls-files` results to take `.gitignore` content into consideration so that you don't copy `target` or other dirs. The limit is only your imagination. https://stackoverflow.com/questions/13713101/rsync-exclude-according-to-gitignore-hgignore-svnignore-like-filter-c

For example I define a function in `.bashrc` like this:

```shell
$ type sync
sync is a function
sync ()
{
    rsync -aP --exclude=.git --exclude=target --exclude=.idea /mnt/e/coding/code/APP ~/projects/APP
}
```
Actually I am grateful that WSL 2 on Win 10 comes out now and it's perfect timing for me. Before it was not possible unless I upgrade to Win 11. I want to thank the dev team for it.
